Serial design draft (#11634)
ARMmbed/mbed-os: docs/design-documents/drivers/serial/serial.md

# Table of contents

1. [Serial design document](#serial-design-document).
1. [Table of contents](#table-of-contents).
    1. [Revision history](#revision-history).
1. [Introduction](#introduction).
    1. [Overview and background](#overview-and-background).
    1. [Requirements and assumptions](#requirements-and-assumptions).
1. [System architecture and high-level design](#system-architecture-and-high-level-design).
    1. [System architecture and component interaction](#system-architecture-and-component-interaction).
1. [Detailed design](#detailed-design).
1. [Usage scenarios and examples](#usage-scenarios-and-examples).
1. [Other Work](#other-work).
    1. [Related work-Simplification of retarget code](#related-work-simplification-of-retarget-code).
    1. [Future Work-Deprecation of Stream](#future-work-deprecation-of-stream).
1. [Roll-out Plan](#roll-out-plan).
1. [Tools and configuration changes](#tools-and-configuration-changes).
1. [Other information](#other-information).
    1. [Deprecations](#deprecations).
    1. [References](#references).

### Revision history

1.0 - Initial version - Evelyne Donnaes - 04/10/2019

# Introduction

### Overview and background

> Mbed OS contains multiple serial classes and multiple ways of printing information to the console. Serial classes have evolved over the years to add new functionality to existing classes and compensate for limitations. The result is that there are now multiple variants, not always well documented; therefore it is not clear which class should be used for what purpose. Furthermore, some of the classes pull substantial dependencies whose benefits are not very clear, for example standard library like objects and abstractions, and parts of the retarget layer.

Reviewer: ```suggestion
...and parts of the system I/O function retarget layer.
```

Author: Fixed in last commit

- It lacks buffering so it cannot be used reliably for input and output without flow control.
- It pulls in the C library stdio system because it uses `Stream`.
- Although it is a `FileHandle`, its implementation is blocking only and has incorrect read semantics.

**UARTSerial**

`UARTSerial` provides buffered I/O. It can be used reliably for input from non-interrupt context and to avoid excess spinning waiting for transmission buffer space. It also does better blocking and correct POSIX-like `FileHandle` semantics. 

It does not use `Stream` so it has the advantage of not pulling in the C stdio library. The drawback is that it doesn't include built-in printf methods, see [Usage scenarios and examples](#usage-scenarios-and-examples) for an example on how to print to a serial port.

**RawSerial**

`RawSerial` provides unbuffered I/O. It has no locks so it is safe to use in interrupts provided that it is the only instance that is using that serial port. Unlike `Serial` and `UARTSerial`, it is not a `FileHandle`.

### Requirements and assumptions

None

# System architecture and high-level design

Applications will have a choice between two public serial classes: 

> - `BufferedSerial` for a normal application

Reviewer: What is the meaning of `normal application`?

Author: I will remove.

Reviewer 2: You still need to say something about why you'd use one or the other. The point is true that `BufferedSerial` should be the default choice, except for specialised cases.

`RawSerial` is something that doesn't even exist in any real operating system - you can't do it from application space at all, and you could only do it from inside the kernel by bypassing the normal drivers. You shouldn't be going near it unless you really know what you're doing.

(And the fact that the console defaults to `DirectSerial` is a pretty horrible legacy thing - should we finally fix that for Mbed OS 6?).

I can see valid reasons to use `RawSerial`, but they're right at the pointy end of absolutely maximising performance and absolutely minimising memory.

Author: We are removing `DirectSerial` now so it will come in 5.15 timeframe, which is a good thing!
You are right; I will rephrase to say that BufferedSerial should be the default choice.

Reviewer 2: I was meaning the console defaulting to buffered serial in 6, rather than whatever the minimal/unbuffered version is called.

Reviewer: `DirectSerial` is the default choice in the re-target code. My understanding is that `MinimalConsole` should be the default choice in the re-target code.
@evedon Why should `BufferedSerial` be the default choice? I thought the default configuration of Mbed OS should be as small as possible.

Author: If an application uses `UnbufferedSerial` then it needs to understand the internals of serial in order to get expected I/O behaviour.

Reviewer: Was [that](https://github.com/ARMmbed/mbed-os/pull/11634#discussion_r341530084) to answer my question?

Author: Unbuffered I/O involves reading or writing data one element at a time. The application needs to understand who has access to the serial port so that the writing or reading of characters is as expected. If you are writing from a critical section then you can expect a buffer of characters to be written in sequence; if you are not in a critical section then an unbuffered write could be interrupted if you don't have sole access to the serial port.

Using unbuffered I/O requires more knowledge so it should not be the default case. As written in the design `UnbufferedSerial` is provided for specialised cases like interrupt driven applications or applications that need to control access to the serial port.

We aim to provide a small memory footprint but the goal of the default configuration & API is easy of use and simplicity.

Reviewer: My question was about the re-target code. Which class should the re-target code use by default? In other words, what should the default value of `platform.stdio-minimal-console-only` be? (`"true"` = `MinimalConsole` is default, `"false"` = `UnbufferedSerial` is default)

Author: Sorry I got confused between the two different points. I was answering why BufferedSerial should be the default choice for an application.

The new internal MinimalConsole is the implementation when `stdio-minimal-console-only` is true. When it is false then we have:

```
static FileHandle *default_console()
{
#if MBED_CONF_TARGET_CONSOLE_UART && DEVICE_SERIAL
#  if MBED_CONF_PLATFORM_STDIO_BUFFERED_SERIAL
    static UARTSerial console(STDIO_UART_TX, STDIO_UART_RX, MBED_CONF_PLATFORM_STDIO_BAUD_RATE);
#   if   CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_RTS
    console.set_flow_control(SerialBase::RTS, STDIO_UART_RTS, NC);
#   elif CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_CTS
    console.set_flow_control(SerialBase::CTS, NC, STDIO_UART_CTS);
#   elif CONSOLE_FLOWCONTROL == CONSOLE_FLOWCONTROL_RTSCTS
    console.set_flow_control(SerialBase::RTSCTS, STDIO_UART_RTS, STDIO_UART_CTS);
#   endif
#  else
    static DirectSerial console(STDIO_UART_TX, STDIO_UART_RX, MBED_CONF_PLATFORM_STDIO_BAUD_RATE);
#  endif
#else // MBED_CONF_TARGET_CONSOLE_UART && DEVICE_SERIAL
    static Sink console;
#endif
    return &console;
}
```
If `MBED_CONF_PLATFORM_STDIO_BUFFERED_SERIAL` is true then BufferedSerial (UARTSerial) is selected. Otherwise we will use `UnbufferedSerial` instead of `DirectSerial`.

So, to answer your question, `platform.stdio-minimal-console-only` ? ("true" = MinimalConsole, "false" = UnbufferedSerial or BufferedSerial depending on `stdio-buffered-serial` is default)

Reviewer: That actually did not answer my question 😀.

I have concluded that the default value of `platform.stdio-minimal-console-only` should be `false` as existing apps and tests rely on the APIs that are removed when it is set to `true`. Users have to enable `MinimalConsole` usage in the re-target code as it will be using `UnbufferedSerial\BufferedSerial` by default.

Reviewer 2: > I thought the default configuration of Mbed OS should be as small as possible.

The general current pattern, that I don't think we're changing, is that full functionality is the default; so all the "minimal"s and "bare metals" are off by default.

We only configure "small" where it doesn't impact general functionality. Basically "optimise for space rather than speed" (eg minimise CRC tables), "or leave out unused stuff where possible"; but not "make features unavailable to save space", without explicit consent.

The logic of the existence and documentation of `BufferedSerial` suggests that it should be the default. The only two reasons I know of for it not to be are:

1. reducing memory footprint -- which would fall under the "minimal" default we just said we don't do
1. backwards compatibility with Mbed OS 2/3/5.

The backwards compatibility issues I'm aware of are:

1. Old-school attempts to separately declare and manipulate `Serial pc(USBTX, USBRX)` will likely fail if the console is using `BufferedSerial` - you get away with it with `UnbufferedSerial`/`DirectSerial`. This can hit either simple apps that can be simply changed to use `printf`, or complex ones that juggle console output with their own IRQ input. The latter would need re-engineering, or would have to manually set buffered console to false.
1. Default power consuption is high because reception IRQ is enabled; anyone wanting to enter deep sleep needs to do `disable_input`. (Maybe there could be a JSON setting to have the retarget do that for you?)

- `UnbufferedSerial` for an interrupt driven application or one that needs to have more control

### System architecture and component interaction

> The below diagram shows the inheritance hierarchy for the serial classes.

Reviewer: Is this diagram slightly incorrect? It does not only show serial classes, it also includes `FileHandle`.

Author: The diagram is correct since I want to show that BufferedSerial & UnbufferedSerial inherit from both SerialBase **and** Filehandle.

Reviewer: We have diagrams after that show more of the bases classes. Was this one necessary?

Author: This diagram goes in less details than the other ones. It's true that now that SerialMinimal has been removed, there is a lot of similarities between the general diagram and the detailed ones. Nevertheless I like to see diagrams in designs and I'd prefer to keep it.

Reviewer: Ok. I do not have the button to resolve the comment.

![Serial classes](./diagrams/serial-classes.png)

# Detailed design

### Detailed design : BufferedSerial

> `BufferedSerial` is `UARTSerial` renamed to convey the original purpose of the class. In addition, the new class will allow configurability of the transmit and receive buffer size so that each instance of the class can configure its buffer size according to its intended usage, for example a modem may need a large input buffer and a console may not need any input buffer.

Reviewer 3: The name reminds  me earlier component available on developer page. It's been used, I recall the name UARTSerial was chosen not to create a naming clash.
I assume they are not compatible and users would need to update their apps if they move to this new class? Might be just noted here

Author: Indeed I can see references to a BufferedSerial class if I google it. But I don't think the class was even part of a Mbed OS release. Wouldn't it create even more confusion if we refer to a class not in Mbed OS source code?

As part of this work, we will deprecate Serial and RawSerial so users will get warnings if they use deprecated code (The BufferedSerial class publicly derived from RawSerial).

Reviewer 3: It has been used mainly in mbed 2 apps. Correct, was not part of the release.  Just wanted to make sure we are aware. It shall be fine as it is.

The below diagram shows the detailed inheritance hierarchy for `BufferedSerial`.

![BufferedSerial class](./diagrams/buffered-serial.png)

```C
class BufferedSerial : private SerialBase, public FileHandle, private NonCopyable<BufferedSerial>
```

`BufferedSerial` privately inherits from `SerialBase`. It means that applications are forced to use the transmit and receive buffers since they cannot access `SerialBase` member functions directly.

**API description**

These are the new constructors for `BufferedSerial`. The first constructor creates the buffers dynamically.

```C
    /** Create a BufferedSerial port, connected to the specified transmit and receive pins, with a particular baud rate.
     *  @param tx Transmit pin
     *  @param rx Receive pin
     *  @param baud The baud rate of the serial port (optional, defaults to MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE)
     *  @param rx_buffer_size The size of the receive buffer (optional, defaults to MBED_CONF_DRIVERS_UART_SERIAL_RXBUF_SIZE)
     *  @param tx_buffer_size The size of the transmit buffer (optional, defaults to MBED_CONF_DRIVERS_UART_SERIAL_TXBUF_SIZE)
     */

>     BufferedSerial(PinName tx,

Reviewer 3: unbufferedserial contains a note about tx/rx being NC, not valid for this class?

Author: The comment should equally apply to BufferedSerial. I will add a note.

public:
    /** Create a UnbufferedSerial port, connected to the specified transmit and receive pins, with the specified baud.
     *
     *  @param tx Transmit pin
     *  @param rx Receive pin
     *  @param baud The baud rate of the serial port (optional, defaults to MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE)
     *
     *  @note
     *    Either tx or rx may be specified as NC if unused
     */
    UnbufferedSerial(PinName tx, PinName rx, int baud = MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE);

    virtual ssize_t write(const void *buffer, size_t size);
    virtual ssize_t read(void *buffer, size_t size);
    virtual off_t seek(off_t offset, int whence = SEEK_SET);
    virtual off_t size();
    virtual int isatty();
    virtual int close();
    virtual short poll(short events) const;

#if !(DOXYGEN_ONLY)
protected:

    /* Acquire exclusive access to this serial port
     */
    virtual void lock(void);

    /* Release exclusive access to this serial port
     */
    virtual void unlock(void);
#endif
};
```

# Usage scenarios and examples

### Scenario 1 `Printing to stdout`

> There is no change in the current APi in regard to printing to stdout.  Applications should call `printf` directly as shown below. The console baud rate can be configured via `platform.stdio-baud-rate` in `mbed_app.json`.

Reviewer: ```suggestion
There is no change in the current APi in regard to printing to stdout.  Applications should call `printf` directly as shown below. The console default baud rate can be configured via `platform.stdio-baud-rate` in `mbed_app.json`.
```

Reviewer 2: There's currently no direct C++ API to change the console baud rate, aside from that start-up setting, so I think it's just the baud rate.

For comparison there is the other setting `platform.default-serial-baud-rate`, but that is a true default - that's the speed for a `Serial` object where it's never set, and it's perfectly normal to use the real API, rather than that default.

Author: Thanks for the clarification.

Reviewer: I didn't realise there was `platform.stdio-baud-rate` as well as `platform.default-serial-baud-rate`. I was only aware of the latter and thought it was the one in the text.

Reviewer 2: I don't really see the value of `platform.default-serial-baud-rate`. It's a global setting applying to all  serial ports. I struggle to think of a use case - any device with multiple ports is going to want each one (GPS, modem, console) individually set, and code for each of them will be specifying baud rate directly anyway (with settings like `platform.stdio-baud-rate`).

Personally, I would have just had default baud rate be fixed 9600. Just like SPI defaults to 8 bit mode 0, 1MHz - there's no config setting for that. Or indeed like the serial format defaults to 8N1 with no config setting.

I'd be inclined to kill that setting just to reduce clutter and confusion.

Just wondering how to deprecate settings. Maybe change the default to `null`, then have the code do:

    #ifdef MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE
    #define DEFAULT_SERIAL_BAUD_RATE MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE
    #warning "Deprecated blah blah"
    #else
    #define DEFAULT_SERIAL_BAUD_RATE 9600
    #endif

> There is no change in the current APi in regard to printing to stdout.  Applications should call `printf` directly as shown below. The console baud rate can be configured via `platform.stdio-baud-rate` in `mbed_app.json`.

Reviewer 4: I don't really like the `should`, I would say `can` and also add a small disclaimer, something like `(it can negatively impact the image size)`

Reviewer 2: Negatively impact compared to what? I believe this section started as a comparator to `Serial::printf`, and I think that's its focus. The example doesn't even need to be `printf`, it could have been `puts`.

If we want to talk about `printf` versus alternatives, that's at a bit of a tangent to the serial work, but I guess the possibilities are:

* Using `printf`
* Using `printf` with `minimal-printf` on, reducing formatter
* Only using `puts` (or all your `printf`s are simple enough for compiler to turn them into `puts`), avoiding formatter
* Using just `write`, avoiding C library altogether
* Turning on `minimal-console` (in conjunction with any of the 4 previous options)

Reviewer 4: I meant calling stdlib printf from the application, and it's a good point it's not a place to talk about memory optimisations. But I don't want to push people to printf too much.

Author: I changed `should` into `can` and put the example using `write` first. Discussion related to printf comes after so I am not emphasing the use of printf.

```C
printf("Hello");
```
or
```C
fprintf(stdout, "Hello");
```

### Scenario 2 `Printing to a serial port other than stdout`

The below example shows pseudocode for how to print `Hello` to `serial` port using the new API. `serial.printf` cannot be called directly; instead the application first calls _fdopen_ to get a _FILE*_ to use the printf function.

> ```C

Reviewer 4: Could we flip the order and give the most "straightforward" example first? That is the direct usage of the class and write. Also mentioning why would you choose to use the file way (I'm assuming redirection and extended formating). How does it affect the image size?

Author: Agreed.

# Table of contents

1. [Serial design document](#serial-design-document).
1. [Table of contents](#table-of-contents).
    1. [Revision history](#revision-history).
1. [Introduction](#introduction).
    1. [Overview and background](#overview-and-background).
    1. [Requirements and assumptions](#requirements-and-assumptions).
1. [System architecture and high-level design](#system-architecture-and-high-level-design).
    1. [System architecture and component interaction](#system-architecture-and-component-interaction).
1. [Detailed design](#detailed-design).
1. [Usage scenarios and examples](#usage-scenarios-and-examples).
1. [Other Work](#other-work).
    1. [Related work-Simplification of retarget code](#related-work-simplification-of-retarget-code).
    1. [Future Work-Deprecation of Stream](#future-work-deprecation-of-stream).
1. [Roll-out Plan](#roll-out-plan).
1. [Tools and configuration changes](#tools-and-configuration-changes).
1. [Other information](#other-information).
    1. [Deprecations](#deprecations).
    1. [References](#references).

### Revision history

1.0 - Initial version - Evelyne Donnaes - 04/10/2019

# Introduction

### Overview and background

> Mbed OS contains multiple serial classes and multiple ways of printing information to the console. Serial classes have evolved over the years to add new functinality to existing classes and compensate for limitations. The result is that there are now multiple variants, not always well documented; therefore it is not clear which class shoud be used for what purpose. Futhermore, some of the classes pull substantial dependencies whose benefits are not very clear, for example standard library like objects and abstractions, and parts of the retarget layer.

Reviewer: Typo corrections.
```suggestion
Mbed OS contains multiple serial classes and multiple ways of printing information to the console. Serial classes have evolved over the years to add new functionality to existing classes and compensate for limitations. The result is that there are now multiple variants, not always well documented; therefore it is not clear which class should be used for what purpose. Furthermore, some of the classes pull substantial dependencies whose benefits are not very clear, for example standard library like objects and abstractions, and parts of the retarget layer.
```

This document collects proposals made by various contributors (see [references](#references)) and presents a simplified serial class hierarchy. The aim is to offer two public classes, one for unbuffered I/O and the other for buffered I/O, with a clear definition of which one should be used when. All other serial classes are deprecated and should be removed at a later stage. In addition, the serial classes dependencies will be optimised to only include components that provide valuable functionality.

#### Serial classes in Mbed OS 5.14

**Serial**

`Serial` provides unbuffered I/O and is using blocking HAL calls. It only works for simple printing. 

Its main limitations are:
- It uses a mutex lock so it cannot be used from interrupts. 
- It wastes a lot of CPU time spinning waiting for serial output and it cannot be used reliably for input because it needs interrupt-speed response on read calls but its high-level API cannot be used from interrupts.

> - It lacks buffering so it cannot be used reliably for input and output without flow control.

Reviewer 2: Another flaw is that although it is a `FileHandle`, it's limited (blocking only, and incorrect read semantics). You mention it in `UARTSerial`, but should be here as a limitation.

- It pulls in the C library stdio system because it uses `Stream`.

**UARTSerial**

`UARTSerial` provides buffered I/O. It can be used reliably for input from non-interrupt context and to avoid excess spinning waiting for transmission buffer space. It also does better blocking and correct POSIX-like `FileHandle` semantics. 

> It does not use `Stream` so it has the advantage of not pulling in the C stdio library. The drawback is that it loses the ability to call printf directly.

Reviewer 2: Clarify - it doesn't include built-in `printf` methods, I think you mean

Author: Correct. Will clarify

**RawSerial**

> `RawSerial` provides unbuffered I/O. It has no locks so it is safe to use in interrupts provided that it is the only instance that is using that serial port.

Reviewer 2: Should note that unlike `Serial` and `UARTSerial`, it is not a `FileHandle`.  (Hence current the internal need for `DirectSerial`, which is).

Author: Indeed

### Requirements and assumptions

None

# System architecture and high-level design

Applications will have a choice between two public serial classes: 
- `BufferedSerial` for a normal application
- `UnbufferedSerial` for an interrupt driven application or one that needs to have more control

### System architecture and component interaction

The below diagram shows the inheritance hierarchy for the serial classes.

![Serial classes](./diagrams/serial-classes.png)

> `MinimalSerial` is an abstract class that extracts basics I/O functionality from `SerialBase` so it can be re-used for a minimal serial console in the retarget code.

Reviewer 2: I'm a bit doubtful about the value of this, as a separation, but will reserve judgement until I see it.

Author: Since this is what is used by the minimal console in the simplified retarget code, it might save a bit of memory and make the MinimalConsole implementation simpler. But if this is not the case, we will revert.

Reviewer: Using `DirectSerial`
```
Total Static RAM memory (data + bss): 205934 bytes
Total Flash memory (text + data): 46299 bytes
```

The below are with `MinimalConsole` inheriting from `MinimalSerial`
```
Total Static RAM memory (data + bss): 205978 bytes
Total Flash memory (text + data): 44352 bytes
```

The below are with `MinimalConsole` without inheritance
```
Total Static RAM memory (data + bss): 205782(-196) bytes
Total Flash memory (text + data): 44312(-40) bytes
```

Author: MinimalSerial is not bringing memory ssaving so it is removed.

# Detailed design

### Detailed design : BufferedSerial

`BufferedSerial` is `UARTSerial` renamed to convey the original purpose of the class. In addition, the new class will allow configurability of the transmit and receive buffer size so that each instance of the class can configure its buffer size according to its intended usage, for example a modem may need a large input buffer and a console may not need any input buffer. 

The below diagram shows the detailed inheritance hierarchy for `BufferedSerial`.

> ![BufferedSerial class](./diagrams/buffered-serial.png)

Reviewer 2: I'd suggest annotating the `BufferedSerial`->`SerialBase` inheritance somehow to note that it's private. As `SerialBase` isn't polymorphic, that forces people to via the buffer. If the `SerialBase` was public, people could bypass the buffer, which would be weird.

**API description**

This is the new constructor for `BufferedSerial` to pass in the serial buffers size.

```C
    /** Create a BufferedSerial port, connected to the specified transmit and receive pins, with a particular baud rate.
     *  @param tx Transmit pin
     *  @param rx Receive pin
     *  @param baud The baud rate of the serial port (optional, defaults to MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE)
     *  @param rx_buffer_size the size of the receive buffer (optional, defaults to MBED_CONF_DRIVERS_UART_SERIAL_RXBUF_SIZE)
     *  @param tx_buffer_size the size of the transmit buffer (optional, defaults to MBED_CONF_DRIVERS_UART_SERIAL_TXBUF_SIZE)
     */
    BufferedSerial(PinName tx, 
                   PinName rx, 
                   int baud = MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE, 

>                    uint32_t rx_buffer_size = MBED_CONF_DRIVERS_UART_SERIAL_RXBUF_SIZE,

Reviewer 2: As @bulislaw has reminded me a couple of times, we mustn't force dynamic allocation. This one would need to dynamically allocate the buffers, but we can also have constructors that pass in a buffer pointer:

    BufferedSerial(PinName tx, 
                   PinName rx, 
                   int baud = MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE, 
                   void *rx_buffer,
                   size_t rx_buffer_size,
                   void *tx_buffer,
                   size_t tx_buffer_size);

(Note there's no reason not to use size_t).

Also note that the recent pinmap stuff will multiply up the number of constructors...

Author: Good point.

The below diagram shows the detailed inheritance hierarchy for `UnbufferedSerial`.

![UnbufferedSerial class](./diagrams/unbuffered-serial.png)

**API description**

This is the proposed API for the `UnbufferedSerial` class.

```C
class UnbufferedSerial: public SerialBase, public FileHandle, private NonCopyable<UnbufferedSerial> {

public:
    /** Create a UnbufferedSerial port, connected to the specified transmit and receive pins, with the specified baud.
     *
     *  @param tx Transmit pin
     *  @param rx Receive pin
     *  @param baud The baud rate of the serial port (optional, defaults to MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE)
     *
     *  @note
     *    Either tx or rx may be specified as NC if unused
     */
    UnbufferedSerial(PinName tx, PinName rx, int baud = MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE);

    /** Write a char to the serial port
     *
     * @param c The char to write
     *
     * @returns The written char or -1 if an error occurred
     */

>     int putc(int c);

Reviewer 2: Do we really want/need `putc`/`getc`/`puts`? If implementing `FileHandle`, `read`/`write` do everything, and not sure you need the extra sugar.

Author: You are right. We should only keep read and write.

    /** Read a char from the serial port
     *
     * @returns The char read from the serial port
     */
    int getc();

    /** Write a string to the serial port
     *
     * @param str The string to write
     *
     * @returns 0 if the write succeeds, EOF for error
     */
    int puts(const char *str);

>     virtual ssize_t write(const void *buffer, size_t size);

Reviewer 4: I'm assuming these are virtual but not abstract.

Author: Indeed, this is a virtual function with a default implementation.

    virtual ssize_t read(void *buffer, size_t size);
    virtual off_t seek(off_t offset, int whence = SEEK_SET);
    virtual off_t size();
    virtual int isatty();
    virtual int close();
    virtual short poll(short events) const;

#if !(DOXYGEN_ONLY)
protected:

    /* Acquire exclusive access to this serial port
     */
    virtual void lock(void);

    /* Release exclusive access to this serial port
     */
    virtual void unlock(void);
#endif
};
```

> ###  Detailed design : MinimalSerial

Reviewer 4: Worth reiterating that's not going to be a part of public interface.

Author: Will do. Note that this class may go if it does not bring any memory saving to the simplified retarget code.

Author: The class has been removed.

The `MinimalSerial` class is a minimal abstract base class that provides basics I/O functionality. All other serial classes derive from it.

**API description**

This is the proposed API for the `MinimalSerial` class. Only two methods are provided to output `_base_putc()` and receive `_base_getc()` characters on the UART.

```C
class MinimalSerial : private NonCopyable<MinimalSerial> {
protected:
    MinimalSerial(PinName tx, PinName rx, int baud);
    virtual ~MinimalSerial();
    int _base_getc();
    int _base_putc(int c);
};
```

###  Detailed design : MinimalConsole

> `MinimalConsole` is an internal concrete class that derives from `MinimalSerial` and provides basic console functionality that can be used by the retarget code for resources constrained targets, see [Simplication of retarget code](#simplification-of-retarget-code).

Reviewer 2: I'm not convinced this wants to be tied in with any of the `FileHandle` related stuff. The extra virtual overhead of two more classes doesn't seem to be buying anything. And in the simplified retarget, nothing needs/wants to be virtual to keep the size down.

And `MinimalConsole` shouldn't be locked into serial anyway, in which case having both `SerialBase` and `MinimalConsole` based on `MinimalSerial` isn't logical.

If offering the minimal console, it should be based on weak functions to allow the possibility to override (even though it's a parallel override mechanism to the FileHandle-based one, there's no reason not to do it, and plenty of similar weak precedent).

Author: > The extra virtual overhead of two more classes doesn't seem to be buying anything. And in the simplified retarget, nothing needs/wants to be virtual to keep the size down.

I think you might be right. If it increases the simplified retarget then it will be removed. I don't have a strong opinion on `MinimalSerial` other than the fact that it provides basic functionality for both a console and `SerialBase`.

Author: MinimalConsole was removed

Reviewer: > MinimalConsole was removed

Was it? when? Or did you mean `MinimalSerial`?

**API description**

```C
class MinimalConsole : public MinimalSerial, private NonCopyable<MinimalConsole> {
public:
    MinimalConsole(PinName tx, PinName rx, int baud);
    virtual ~MinimalConsole();
    int getc();
    int putc(int c);
};
```

The constructor provides arguments to create a serial port connected to the specified transmit and receive pins and set to the specified baud rate. If a target has serial support, by default a serial port is used for the console and the pins and settings for the port selection come from target header files and JSON settings. 

# Usage scenarios and examples

> ### Scenario 1 `Printing to stdout`

Reviewer 4: That's a big change and hit for user experience. I know using printf is not ideal, as it's heavy but it's extremely easy. We just did bunch of work to make it cheaper (minimal printf). I'm not sure how the above proposal affects the default retargeting setup. That being said we should have an option to disable it.

Reviewer: `printf()` will still be available via re-targeting the systems I/O functions as it currently is the case. `UnbufferedSerial` will be used instead of `DirectSerial` and `BufferedSerial` will be used instead of `UARTSerial`.
What is shown here is how to print to the console using the high level function (`fprintf`) if the re-targeting code is not used.

Author: Example corrected and clarified. Printing to stdout is as usual using `printf`. Printing to a serial port requires a call to `fdopen`.

The below example shows pseudocode for how to print `Hello` to stdout using the new API. Printf cannot be called directly; instead the application first calls _fdopen_ to get a _FILE*_ to use the printf function.

```C
BufferedSerial serial(RX, TX, 115200);
FILE *out = fdopen(&serial, "w");
fprintf(out, "Hello");

> ```

Reviewer 2: That's not printing to `stdout`. Printing to `stdout` would be

    printf("Hello");

or

    fprintf(stdout, "Hello");

Users should not be messing with the serial port used by the console - that's bad practice, and exactly one of the things in our examples we should be discouraging.

Previously people did that because it was the only way of setting the console baud rate, but that's no longer true - can be configured via JSON.

This example would be valid for printing to a different serial port.

> Printf cannot be called directly

It absolutely can. I think you mean to say you can't call `serial.printf`?

Author: Thanks for both corrections. On the last point, I did mean to say you can't call `serial.printf`. 
I will add two scenarios: one to print to stdout and one to print to a serial port.

and similarly for `UnbufferedSerial`

```C
UnbufferedSerial serial(RX, TX, 115200);
FILE *out = fdopen(&serial, "w");
fprintf(out, "Hello");
```

# Other Work

## Related work-Simplification of retarget code

In Mbed, retargeting is implemented by redefining the system I/O functions.
The `FileHandle` abstract class provides an interface with file-like operations, such as read and write that can be redefined by the retarget code. The high-level library functions perform input/output by calling the low-level functions which use the system I/O functions to interface with hardware.

The relationship of `FileHandle` and other APIs is as follows:

> ![Retarget](./diagrams/retarget.png)

Reviewer 2: In that diagram should probably also note `FileSystem` alongside `FileHandle`, which provides the `open` call.

> The POSIX layer and the static array of `FileHandle*` allocated by the retarget code for file handling consume a lot of memory. A constrained target that only uses a console does not need file handling and should be able to turn this functionality off.

Reviewer 2: Also the virtual-ness of the FileHandle means ROM.

 A new configuration parameter is introduced to compile the file handling code out. The application can override the default configuration in mbed_app.json as follows:

```C
    "target_overrides": {
        "*": {
            "platform.stdio-minimal-console-only": true
        }
    }
```

There are two approaches to achieve a minimal retarget layer. You can:
1. Redefine the low-level library functions. A detailed description can be found [here](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0475m/chr1358938931411.html)
1. Redefine the system I/O functions.

> A minimal console only needs to write a single character at a time. So redefining the default fputc() to directly write to the serial port if the output stream is stdout and bypassing the system I/O functions should achieve higher memory savings. If we take this approach, we will have to rework some error handlers that rely on the POSIX form of `write(STDOUT_FILENO, buf, len)` to do emergency printing.

Reviewer 2: That method is also toolchain dependent - I don't know if we've established how easy it is with the other 2.

Author: Indeed the first approach looks more complex. We should only pursue if it brings more memory saving. We'll decide during implementation.

- Use `UnbufferedSerial` instead of `DirectSerial`
- Remove `DirectSerial`
- Add a configuration parameter to compile out file handling feature
    - Alternative uses weak console_putc/getc functions

## Future work-Deprecation of Stream

The `Stream` class pulls in the <stdio.h> library. It should be replaced by drivers implementing `FileHandle` directly and applications using `fdopen(FileHandle *)` to get a `FILE*` to use full C/C++ stream features. 

`Stream` is currently used by the following classes:
- `Serial`
- `USBKeyboard`
- `USBMouseKeyboard`
- `USBSerial`

Deprecation of `Stream` should be planned in a future release.

# Roll-out Plan

This feature will be implemented in different phases as follows:

Phase 1:
- [Deprecation of APIs](#Deprecations)
- Implementation of new class `UnbufferedSerial`
- Implementation of new class `BufferedSerial`
- Use `UnbufferedSerial` instead of `DirectSerial` in retarget code

Phase 2:
- Removal of deprecated APIs

# Tools and configuration changes

### Tool changes

None

### Configuration changes

None

# Other information

### Deprecations

The following APIs will be deprecated:
- `Serial` class
- `RawSerial` class

> - `UARTSerial` class

Reviewer 4: DirectSerial?

Reviewer: `DirectSerial` ([here](https://github.com/ARMmbed/mbed-os/blob/master/platform/source/mbed_retarget.cpp#L141).) will not be needed as `UnbufferedSerial` will be used instead.

Reviewer 2: Doesn't need to be deprecated, as it's not public. Can just be removed.
